server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4363,54 +4500,59 @@ assert getActiveOperationsCount() == OPERATIONS_BLOCKED
         assert globalCheckpoint == getLastSyncedGlobalCheckpoint();
         synchronized (engineMutex) {
             verifyNotClosed();
-            // we must create both new read-only engine and new read-write engine under engineMutex to ensure snapshotStoreMetadata,
-            // acquireXXXCommit and close works.
-            final Engine readOnlyEngine = new ReadOnlyEngine(
-                newEngineConfig(replicationTracker),
-                seqNoStats,
-                translogStats,
-                false,
-                Function.identity(),
-                true,
-                false
-            ) {
-                @Override
-                public IndexCommitRef acquireLastIndexCommit(boolean flushFirst) {
-                    synchronized (engineMutex) {
-                        if (newEngineReference.get() == null) {
-                            throw new AlreadyClosedException("engine was closed");
+            engineResetLock.readLock().lock(); // prevent engine resets during rollback
```

Comment:
We seem to already be protected against that here - like in close, perhaps either comment on that. OK to leave in place, seems good for lock ordering if we also take the `mutex` (and thus I do not have to check whether we do).

Reply:
I removed the read lock in 4a8186247796830137ad6cfde5e8aeff0d7c882d.